api/resource_groups_version_service.go

```diff
@@ -0,0 +1,367 @@
+	return &ResourceGroupsVersionService{
+		c,
+		&ResourceGroupsService{c},
+		&ResourceGroupsV2Service{c},
+		&FeatureFlagsService{c},
+	}
+}
+
+// NewResourceGroup returns an instance of the ResourceGroupData struct with the
+// provided ResourceGroup type, name and the props field as an interface{}.
+//
+// NOTE: This function must be used by any ResourceGroup type.
+//
+// Basic usage: Initialize a new ContainerResourceGroup struct, then
+//              use the new instance to do CRUD operations
+//
+//   client, err := api.NewClient("account")
+//   if err != nil {
+//     return err
+//   }
+//
+//   group := api.NewResourceGroup("container resource group",
+//     api.ContainerResourceGroup,
+//     api.ContainerResourceGroupData{
+//       Props: api.ContainerResourceGroupProps{
+//			Description:     "all containers,
+//			ContainerLabels: ContainerResourceGroupAllLabels,
+//			ContainerTags:   ContainerResourceGroupAllTags,
+//		},
+//     },
+//   )
+//
+//   client.V2.ResourceGroups.Create(group)
+//
+func NewResourceGroup(name string, iType ResourceGroupType, props interface{}) ResourceGroupData {
+	return ResourceGroupData{
+		Name:    name,
+		Type:    iType.String(),
+		Enabled: 1,
+		Props:   props,
+	}
+}
+
+// NewResourceGroupWithQuery Only available with RGv2 beta
+func NewResourceGroupWithQuery(name string, iType ResourceGroupType,
+	description string, query *RGQuery) ResourceGroupDataWithQuery {
+	return ResourceGroupDataWithQuery{
+		Name:        name,
+		Type:        iType.String(),
+		Enabled:     1,
+		Query:       query,
+		Description: description,
+	}
+}
+
+func isRGV2FlagEnabled(featureFlagService *FeatureFlagsService) bool {
+	response, err := featureFlagService.GetFeatureFlagsMatchingPrefix("PUBLIC.rgv2.cli")
+
+	if err != nil {
+		return false
+	}
+
+	return len(response.Data.Flags) >= 1
+}
+
+func (svc *ResourceGroupsVersionService) Get(guid string, response interface{}) error {
+	var rawResponse resourceGroupWorkaroundResponse
+	err := svc.get(guid, &rawResponse)
+	if err != nil {
+		return err
+	}
+
+	if rawResponse.Data.Query != nil {
+		return castRGV2WorkAroundResponse(rawResponse, response)
+	} else {
+		return castRGV1WorkAroundResponse(rawResponse, response)
+	}
+}
+
+func (svc *ResourceGroupsVersionService) Create(group ResourceGroupsInterfaceData) (
+	response ResourceGroupResponse,
+	err error,
+) {
+	if group.GetProps() != nil {
+		response, err = svc.v1ResourceGroupService.Create(group.(ResourceGroupData))
+		return
+	}
+
+	isV2FlagEnabled := isRGV2FlagEnabled(svc.featureFlagService)
+
+	if isV2FlagEnabled && group.GetQuery() != nil {
+		createResponse, createErr := svc.v2ResourceGroupService.Create(group.(ResourceGroupDataWithQuery))
+		if createErr != nil {
+			err = createErr
+			return
+		}
+
+		err = castResourceGroupV2Response(createResponse, &response)
+		return
+	}
+
+	err = errors.New("Unable to create resource group")
```

Comment:
Is there any more context we can put in this error. What scenario would this be returned.

In the above code, 
  if props is not nil;then we use the v1 create and return
  if v2 is enabled and the query is not nil; attempt v2 create and cast response and return;
  
This error will return if the input is doesn't have props, doesn't have have v2 flag and a query. With this condition, maybe we can return eariler.

Reply:
Ok I've refactored it a bit.